bpfd-operator/controllers/bpfd-agent/bpfd-agent.go

```diff
@@ -241,147 +251,141 @@ func (r *BpfProgramReconciler) reconcileBpfProgramConfig(ctx context.Context,
+			// owner relationship can take care of cleanup
+			if !BpfProgramConfig.DeletionTimestamp.IsZero() {
+				r.Logger.V(1).Info("bpfProgramConfig is deleted, don't load program, remove finalizer")
+				if controllerutil.ContainsFinalizer(bpfProgram, BpfdAgentFinalizer) {
+					controllerutil.RemoveFinalizer(bpfProgram, BpfdAgentFinalizer)
+					err := r.Update(ctx, bpfProgram)
+					if err != nil {
+						return false, err
+					}
+				}
 
-	// Compare the desired state to existing bpfd state
-	v, ok := nodeState[programKey]
-	// bpfProgram doesn't exist on node
-	if !ok {
-		r.Logger.V(1).Info("bpfProgram doesn't exist on node")
+				return false, nil
+			}
+
+			// Make sure if we're not selected just exit
+			if !isNodeSelected {
+				r.Logger.V(1).Info("bpfProgramConfig does not select this node")
+				// Write NodeNodeSelected status
+				r.updateStatus(ctx, bpfProgram, BpfProgCondNotSelected)
 
-		// If BpfProgramConfig is being deleted just remove agent finalizer so the
-		// owner relationship can take care of cleanup
-		if !BpfProgramConfig.DeletionTimestamp.IsZero() {
-			r.Logger.V(1).Info("bpfProgramConfig is deleted, don't load program, remove finalizer")
+				return false, nil
+			}
+
+			// otherwise load it
+			loadRequest, err := internal.BuildBpfdLoadRequest(modBpfProgramConfig)
+			if err != nil {
+				r.updateStatus(ctx, bpfProgram, BpfProgCondNotLoaded)
+
+				return true, fmt.Errorf("failed to generate bpfd load request: %v",
+					err)
+			}
+			r.Logger.Info("Loading bpfProgram", "loadRequest", loadRequest)
+			return r.loadBpfdProgram(ctx, loadRequest, bpfProgram)
+		}
+
+		// BpfProgram exists but either BpfProgramConfig is being deleted or node is no
+		// longer selected....unload program
+		if !BpfProgramConfig.DeletionTimestamp.IsZero() || !isNodeSelected {
+			r.Logger.V(1).Info("bpfProgram exists on Node but is scheduled for deletion or node is no longer selected", "isDeleted", !BpfProgramConfig.DeletionTimestamp.IsZero(),
+				"isSelected", isNodeSelected)
 			if controllerutil.ContainsFinalizer(bpfProgram, BpfdAgentFinalizer) {
+				if retry, err := r.unloadBpfdProgram(ctx, bpfProgram); err != nil {
+					return retry, err
+				}
+
+				// Remove bpfd-agentFinalizer. Once all finalizers have been
+				// removed, the object will be deleted.
 				controllerutil.RemoveFinalizer(bpfProgram, BpfdAgentFinalizer)
 				err := r.Update(ctx, bpfProgram)
 				if err != nil {
 					return false, err
 				}
-			}
 
-			return false, nil
-		}
-
-		// Make sure if we're not selected just exit
-		if !isNodeSelected {
-			r.Logger.V(1).Info("bpfProgramConfig does not select this node")
-			// Write NodeNodeSelected status
-			r.updateStatus(ctx, bpfProgram, BpfProgCondNotSelected)
+				// If K8s hasn't cleaned up here it means we're no longer selected
+				// write NodeNodeSelected status ignoring error (object may not exist)
+				r.updateStatus(ctx, bpfProgram, BpfProgCondNotSelected)
```

Comment:
Can you make this into a function that we return? i.e we should ALWAYS exit out of the reconcile loop if we end up changing the status